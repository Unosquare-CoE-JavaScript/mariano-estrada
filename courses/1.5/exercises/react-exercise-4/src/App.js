import React, { useContext } from 'react';

import Login from './components/Login/Login';
import Home from './components/Home/Home';
import MainHeader from './components/MainHeader/MainHeader';
import AuthContext from './context/auth-context';
// import AuthContext from "./context/auth-context";

function App() {
  const ctx = useContext(AuthContext);
  // const [isLoggedIn, setIsLoggedIn] = useState(false);

  // useEffect(() => {
  //   const storedUserLoggedIn = localStorage.getItem("isLoggedIn");

  //   if (storedUserLoggedIn === 1) {
  //     setIsLoggedIn(true);
  //   }
  // }, []);

  // const loginHandler = (email, password) => {
  //   // We should of course check email and password
  //   // But it's just a dummy/ demo anyways

  //   setIsLoggedIn(true);
  // };

  // const logoutHandler = () => {

  //   setIsLoggedIn(false);
  // };

  return (
    <>
      <MainHeader />
      <main>
        {!ctx.isLoggedIn && <Login />}
        {ctx.isLoggedIn && <Home />}
      </main>
    </>
  );
}

export default App;
