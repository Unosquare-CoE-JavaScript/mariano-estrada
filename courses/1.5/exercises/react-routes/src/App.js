import React, { Suspense } from 'react';
import { Route, Routes, Navigate, Link } from 'react-router-dom';

import AllQuotes from './pages/AllQuotes';
import QuoteDetail from './pages/QuoteDetail';
import NotFound from './pages/NotFound';
import Layout from './components/layout/Layout';
import Comments from './components/comments/Comments';

const NewQuote = React.lazy(() => import('./pages/NewQuote'));

function App() {
  return (
    <Layout>
      <Suspense fallback={<div className="centered">loading</div>}>
        <Routes>
          <Route path="/" element={<Navigate replace to="/quotes" />} />
          <Route path="/quotes" element={<AllQuotes />} />
          <Route path="/quotes/:quoteId" element={<QuoteDetail />}>
            <Route
              path=""
              element={
                <div className="centered">
                  <Link className="btn--flat" to={`comments`}>
                    Load Comments
                  </Link>
                </div>
              }
            />
            <Route path={`comments`} element={<Comments />} />
          </Route>
          <Route path="/new-quote" element={<NewQuote />} />
          <Route path="*" element={<NotFound />} />
        </Routes>
      </Suspense>
    </Layout>
  );
}

export default App;
