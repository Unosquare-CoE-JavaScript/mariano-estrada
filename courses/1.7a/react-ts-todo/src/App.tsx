import React, {useState} from 'react'
import TodoList from './Components/TodoList'
import NewTodo from './Components/NewTodo'
import { Todo } from './todo.model'

const App: React.FC = () => {
  const [todos, setTodos] = useState<Todo[]>([])

  const todoAddHandler = (text: string)=>{
      setTodos(prevTodos => [...todos, {id: Math.random().toString(), text: text}])
  }

  const deleteTodoHandler = (todoId: string) =>{
    setTodos(prevTodos =>{
      return prevTodos.filter(todo => todo.id !== todo.id)
    })
  } 
  return (
    <div>
      <NewTodo onAddTodo={todoAddHandler} />
      <TodoList items={todos} onDeleteTodo={deleteTodoHandler}/>
    </div>
  )
}

export default App
