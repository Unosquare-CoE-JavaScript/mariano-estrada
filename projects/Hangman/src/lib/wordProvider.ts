export type WordProvider = () => string;

const words = [
  'React',
  'Typescript',
  'Web',
  'Javascript',
  'Vue',
  'Unosquare',
  'Array',
  'String',
  'Java',
  'Localhost',
  'Developer',
];

export function makeWordProvider() {
  if (process.env.NODE_ENV === 'test') {
    return () => 'vue';
  } else {
    return () => {
      const newRandomIndex = Math.floor(Math.random() * words.length);
      return words[newRandomIndex].toLowerCase();
    };
  }
}
