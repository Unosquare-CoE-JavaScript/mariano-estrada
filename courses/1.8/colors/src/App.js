import { useState } from 'react'
import './App.css'

function App () {
  const [buttonColor, setButtonColor] = useState('red')
  const [disabled, setDisabled] = useState(false)
  const newbuttonColor = buttonColor === 'red' ? 'blue' : 'red'

  return (
    <>
      <button
        style={{ backgroundColor: buttonColor }}
        onClick={() => setButtonColor(newbuttonColor)}
        disabled={disabled}
      >
        Change to {newbuttonColor}
      </button>
      <input type='checkbox' 
      defaultChecked={disabled}
      onChange={(e)=> setDisabled(e.target.checked)}
      />
    </>
  )
}

export default App
