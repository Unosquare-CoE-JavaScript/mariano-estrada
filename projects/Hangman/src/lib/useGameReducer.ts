import { useReducer } from 'react';
import { WordProvider } from './wordProvider.ts';

export default function useGameReducer(wordProvider: WordProvider) {
    return useReducer(getGameReducer(wordProvider), initialState);
}

export interface GameState {
    guess: number;
    selectedInitialWord: string;
    displayedWord: string;
    guessedLetter: string;
    hasWon: boolean;
    wrongGuess: string[]
}

export type Action =
    | { type: 'RESET_GAME' }
    | { type: 'SUBMIT_GUESS'; guessedLetter: string }
    | { type: 'UPDATE_DISPLAYED_WORD'; updatedDisplayedWord: string }
    | { type: 'SET_GUESSED_LETTER'; guessedLetter: string }
    | { type: 'SET_SELECTED_INITIAL_WORD'; selectedInitialWord: string };

const initialState: GameState = {
    guess: 5,
    selectedInitialWord: '',
    displayedWord: '',
    guessedLetter: '',
    hasWon: false,
    wrongGuess:[]
};

function getGameReducer(wordProvider: WordProvider) {
    return function gameReducer(state: GameState, action: Action): GameState {
        switch (action.type) {
            case 'RESET_GAME': {
                const newSelectedWord = wordProvider();

                return {
                    ...initialState,
                    selectedInitialWord: newSelectedWord,
                    displayedWord: '_ '.repeat(newSelectedWord.length),
                };
            }

            case 'SUBMIT_GUESS': {
                const newSelectedWord = wordProvider();
                const newDisplayedWord = state.displayedWord.split(' ');
                // Actualizar displayedWord
                const updatedDisplayedWord = state.selectedInitialWord
                  .split('')
                  .map((letter, index) =>
                    letter === action.guessedLetter
                      ? action.guessedLetter
                      : newDisplayedWord[index]
                  )
                  .join(' ');
              
                // Quitar espacios de la palabra
                const word = updatedDisplayedWord.split(/\s+/).join('');
                let newGuess = state.guess;
                let newWrongGuesses = [...state.wrongGuess]; 
              
                if (word === state.selectedInitialWord) {
                  return {
                    ...state,
                    displayedWord: updatedDisplayedWord,
                    guessedLetter: action.guessedLetter,
                    guess: newGuess,
                    hasWon: true,
                  };
                }
              
                if (state.selectedInitialWord.indexOf(action.guessedLetter) === -1) {
                  newGuess--;
              
                  if (!newWrongGuesses.includes(action.guessedLetter)) {
                    newWrongGuesses.push(action.guessedLetter);
                    console.log('Wrong Guesses:', newWrongGuesses);
                  }
              

                  if (newGuess === 0) {
                    return {
                      ...initialState,
                      selectedInitialWord: newSelectedWord,
                      displayedWord: '_ '.repeat(newSelectedWord.length),
                    };
                  }
                }
              
                return {
                  ...state,
                  displayedWord: updatedDisplayedWord,
                  guessedLetter: action.guessedLetter,
                  guess: newGuess,
                  hasWon: false,
                  wrongGuess: newWrongGuesses, 
                };
              }


            case 'UPDATE_DISPLAYED_WORD':
                return {
                    ...state,
                    displayedWord: action.updatedDisplayedWord,
                };

            case 'SET_GUESSED_LETTER':
                return {
                    ...state,
                    guessedLetter: action.guessedLetter,
                };

            case 'SET_SELECTED_INITIAL_WORD':
                return {
                    ...state,
                    selectedInitialWord: action.selectedInitialWord,
                };

            default:
                return state;
        }
    };
}