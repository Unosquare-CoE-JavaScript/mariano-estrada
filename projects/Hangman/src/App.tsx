import { useEffect } from 'react';
import './App.css';
import useGameReducer from './lib/useGameReducer.ts';
import { GameStatus } from './components/GameStatus.tsx';
import { HangmanGame } from './components/HangmanGame.tsx';
import { WordProvider } from './lib/wordProvider.ts';

interface AppProps {
  wordProvider: WordProvider;
  selectedWordIndex: number; 
}

function App({ wordProvider }: AppProps) {
  const [state, dispatch] = useGameReducer(wordProvider);

  useEffect(() => {
    dispatch({ type: 'RESET_GAME' });
  }, [dispatch]);

  function startNewGame() {
    dispatch({ type: 'RESET_GAME' });
  }

  function submitGuess() {
    dispatch({ type: 'SUBMIT_GUESS', guessedLetter: state.guessedLetter });
    console.log('submit')
  }

  const { hasWon } = state;

  return (
    <div className="flex flex-col items-center mt-10" data-testid='hangman-container'>
      <div className="text-5xl font-bold mb-10">Hangman Game</div>
      <div className="flex flex-col md:flex-row w-full md:w-4/6 mx-auto justify-center items-center md:space-x-8">
        <div className="flex-auto w-full md:w-2/5 md:order-1 md:mr-4 text-center md:text-left">
          <GameStatus
            guessesLeft={state.guess}
            wrongGuesses={[]}
            onNewGame={startNewGame}
            wrongGuess={state.wrongGuess}
          />
        </div>
        <div className="flex-auto w-full md:w-2/8 md:order-2 md:mt-7 mt-4 flex-col text-center md:text-left">
          <HangmanGame
            displayedWord={state.displayedWord}
            guessedLetter={state.guessedLetter}
            setGuessedLetter={(guessedLetter) =>
              dispatch({
                type: 'SET_GUESSED_LETTER',
                guessedLetter,
              })
            }
            onSubmitGuess={submitGuess}
          />
        </div>
      </div>
      {hasWon && (
        <div className="text-3xl text-green-600 font-bold mt-4">
         You Won! Start a new game
        </div>
      )}
    </div>
  );
}

export default App;