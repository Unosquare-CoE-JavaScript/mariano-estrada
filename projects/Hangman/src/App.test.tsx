import React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import App from './App'
import { makeWordProvider } from './lib/wordProvider'

describe('<App />', () => {
  it('should render without crashing', () => {
    const wordProvider = makeWordProvider()
    const selectedWordIndex = 0
    render(
      <App wordProvider={wordProvider} selectedWordIndex={selectedWordIndex} />
    )
  })
  it('guesses should reduce by 1 when an incorrect letter is submited', async () => {
    // setup
    const wordProvider = makeWordProvider()
    const selectedWordIndex = 0
    render(
      <App wordProvider={wordProvider} selectedWordIndex={selectedWordIndex} />
    )

    // act
    const input = screen.getByLabelText(/Letter/)
    await userEvent.type(input, 'x')

    const button = screen.getByRole('button', { name: 'Submit' })
    await userEvent.click(button)

    // assert
    expect(screen.getByText('Guesses left: 4')).toBeDefined()
    expect(screen.getByText('_ _ _')).toBeDefined()
  })

  it('when there is a wrong guess the wrong letter should appear in the wrong guesses section', async () => {
    // setup
    const wordProvider = makeWordProvider()
    const selectedWordIndex = 0
    render(
      <App wordProvider={wordProvider} selectedWordIndex={selectedWordIndex} />
    )

    // act
    const input = screen.getByLabelText(/Letter/)
    await userEvent.type(input, 'x')

    const button = screen.getByRole('button', { name: 'Submit' })
    await userEvent.click(button)

    // assert
    const wrongGuessElement = screen.getByText('x');
    const hangmanContainer = screen.getByTestId('hangman-container')
    expect(hangmanContainer).toMatchSnapshot()
    expect(wrongGuessElement).toBeDefined();
    expect(screen.getByText('_ _ _')).toBeDefined()
  })


  it('guesses should not reduce by 1 when an correct letter is submited and the correct letter should appear displayed', async () => {
    // setup
    const wordProvider = makeWordProvider()
    const selectedWordIndex = 0
    render(
      <App wordProvider={wordProvider} selectedWordIndex={selectedWordIndex} />
    )

    // act
    const input = screen.getByLabelText(/Letter/)
    await userEvent.type(input, 'v')

    const button = screen.getByRole('button', { name: 'Submit' })
    await userEvent.click(button)

    // assert
    expect(screen.getByText('Guesses left: 5')).toBeDefined()
    expect(screen.getByText('v _ _')).toBeDefined()
  })

  it('a victory modal should appear when the user guesses the word', async () => {
    // setup
    const wordProvider = makeWordProvider()
    const selectedWordIndex = 0
    render(
      <App wordProvider={wordProvider} selectedWordIndex={selectedWordIndex} />
    )

    // act
    const input = screen.getByLabelText(/Letter/)
    await userEvent.type(input, 'v')

    const button = screen.getByRole('button', { name: 'Submit' })
    await userEvent.click(button)

    await userEvent.type(input, 'u')
    await userEvent.click(button)

    await userEvent.type(input, 'e')
    await userEvent.click(button)

    // assert
    expect(screen.getByText('Guesses left: 5')).toBeDefined()
    expect(screen.getByText('v u e')).toBeDefined()
    expect(screen.getByText('You Won! Start a new game')).toBeDefined()
  })

  it('Should reset game when new game is clicked', async () => {
    // setup
    const wordProvider = makeWordProvider()
    const selectedWordIndex = 0
    render(
      <App wordProvider={wordProvider} selectedWordIndex={selectedWordIndex} />
    )

    // act

    const button = screen.getByRole('button', { name: 'New Game' })
    await userEvent.click(button)

    // assert
    expect(screen.getByText('Guesses left: 5')).toBeDefined()
    expect(screen.getByText('_ _ _')).toBeDefined()
  })

  it('Should reset game when the user lost the game', async () => {
    // setup
    const wordProvider = makeWordProvider()
    const selectedWordIndex = 0
    render(
      <App wordProvider={wordProvider} selectedWordIndex={selectedWordIndex} />
    )

    // act

    const input = screen.getByLabelText(/Letter/)
    await userEvent.type(input, 'x')

    const button = screen.getByRole('button', { name: 'Submit' })
    await userEvent.click(button)

    await userEvent.type(input, 'y')
    await userEvent.click(button)

    await userEvent.type(input, 'z')
    await userEvent.click(button)

    
    await userEvent.type(input, 'b')
    await userEvent.click(button)

    
    await userEvent.type(input, 'p')
    await userEvent.click(button)

    // assert
    expect(screen.getByText('Guesses left: 5')).toBeDefined()
    expect(screen.getByText('_ _ _')).toBeDefined()
  })
})
