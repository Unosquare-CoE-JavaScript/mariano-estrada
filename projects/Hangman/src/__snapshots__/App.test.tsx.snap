// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> when there is a wrong guess the wrong letter should appear in the wrong guesses section 1`] = `
<div
  class="flex flex-col items-center mt-10"
  data-testid="hangman-container"
>
  <div
    class="text-5xl font-bold mb-10"
  >
    Hangman Game
  </div>
  <div
    class="flex flex-col md:flex-row w-full md:w-4/6 mx-auto justify-center items-center md:space-x-8"
  >
    <div
      class="flex-auto w-full md:w-2/5 md:order-1 md:mr-4 text-center md:text-left"
    >
      <div
        class="text-2xl font-bold mb-2 mt-5 md:mt-0"
      >
        Guesses left: 
        4
      </div>
      <div
        class="text-2xl font-bold mb-2 mt-5 md:mt-0"
      >
        Wrong Guesses: 
      </div>
      <div
        class="text-1xl font-bold mb-2 mt-3 md:mt-0 flex space-x-2"
      >
        <div
          class="text-lg text-red-500 rounded px-2 py-1"
        >
          x
        </div>
      </div>
      <div
        class="flex justify-center md:justify-start mt-5"
      >
        <button
          class="px-4 py-2 rounded-md font-bold text-sm bg-blue-500 text-white"
        >
          New Game
        </button>
      </div>
    </div>
    <div
      class="flex-auto w-full md:w-2/8 md:order-2 md:mt-7 mt-4 flex-col text-center md:text-left"
    >
      <div
        class="text-6xl"
      >
        _ _ _
      </div>
      <div
        class="text-3xl mt-4"
      >
        <input
          aria-label="Letter"
          class="px-4 py-0 border mt-4 focus:outline-none text-sm rounded-md w-56 h-10"
          placeholder="Type your guess"
          type="text"
          value=""
        />
        <button
          class="px-2 py-2 rounded-md font-bold text-sm bg-blue-500 text-white mt-4"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</div>
`;
